### ISSUE TITLE
Race condition with regards to opening and closing lock when waiting for SMP characteristic

### ISSUE THREAD

**Issue by MarkBorazio** [JuulLabs-OSS/mcumgr-ios#50]
# Issue
I am occasionally seeing a race condition in the `McuMgrBleTransport`, where the lock is closed due to `smpCharacteristic` being `nil` (in `_send(...)`), but then the lock is _immediately_ opened due to the `centralManagerDidUpdateState` function being called to notify for the `.poweredOn` state. This is causing the `_send(...)` function to fail with a `missingCharacteristic` error since there has not been enough time to discover it yet.

I see this when I first instantiate the `FileSystemManager` and attempt to do a read. The `McuMgrBleTransport` has the peripheral, but not the `smpCharacteristic`, so it closes the lock and _should_ call `centralManager.connect(targetPeripheral)`. However, before it can even call `centralManager.connect(targetPeripheral)`, the lock gets opened, and the `lock.block(...)` call that comes after does not block the thread to wait for `smpCharacteristic` to be discovered.

Most of the time this does not happen, since the call to open the lock is made before the call to close it. I am seeing this issue frequently enough to recognise it, however.
For what it's worth, I can confirm that when seeing this issue, the `McuMgrBleTransport` has only been used in a `FileSystemManager` and no other manager.
I have observed this on an iPhone 6, iOS 12.0.0.

# Proposed Solution

As for a solution, I was thinking that perhaps when the lock is closed and opened, a "reason" has to be given. This would prevent the lock from unintentionally opening. For example, we could close the lock like this:
```
lock.close(reason: .waitingForSmpChar)
```
and the way it can successfully open is to open it while providing the same reason:
```
lock.open(reason: .waitingForSmpChar)
```

A call to `open()` with an error will also open the lock regardless of the reason.

If no reason is provided, then any call to `open()` will open the lock.
We may need to consider that the lock can be closed with multiple reasons as well if it is closed multiple times, although looking at the way things are currently implemented, that shouldn't be necessary.

With that all said, I am spitballing here. I have had a look at the `McuMgrBleTransport` and while I somewhat understand it, I am not going to pretend that I know it inside out. I understand if this suggestion is too big of a change.

**Comment by MarkBorazio** [comment]
Just to add more to this:
I quickly threw in some logging statements to the `_send` function in `McuMgrBleTransport` to see what was actually going on. Here is the code:
```
    private func _send<T: McuMgrResponse>(data: Data, callback: @escaping McuMgrCallback<T>) -> Bool {
        // Is Bluetooth operational?
        if centralManager.state == .poweredOff || centralManager.state == .unsupported {
            log(msg: "Central Manager powered off", atLevel: .warning)
            fail(error: McuMgrBleTransportError.centralManagerPoweredOff, callback: callback)
            return false
        }

        // We might not have a peripheral instance yet, if the Central Manager has not
        // reported that it is powered on.
        // Wait until it is ready, and timeout if we do not get a valid peripheral instance
        let targetPeripheral: CBPeripheral

        print("TEST: central manager state 1: \(centralManager.state.rawValue)")
        if let existing = peripheral, centralManager.state != .unknown {
            print("TEST: central manager state 2: \(centralManager.state.rawValue)")
            targetPeripheral = existing
        } else {
            print("TEST: central manager state 3: \(centralManager.state.rawValue)")
            // Close the lock.
           ...
    }
```
I also added print statements before every call to `lock.open()` and `lock.close`, as well as the reason for those calls.

Here is part of what was printed out:
```
TEST: central manager state 1: 0
TEST: central manager state 2: 5
TEST: closing lock: Waiting for SMP Characteristic to be discovered
TEST: opening Lock: Central Manager Powered On
```

You can see that the central manager state changed to `5` (which is `.poweredOn`) before we checked it in that `if` condition, but the `centralManagerDidUpdateState` delegate function that notifies us of this and opens the lock occurred quite a bit later. It was so much later that we had already closed the lock waiting to discover the SMP characteristic, and obviously this is where the issue manifests.

---

I have already forked this and implemented a fix for the issue using my own suggestion in the post above. From my brief testing (using my own app), things worked perfectly. I can create a PR for it if you want.

**Comment by bgiori** [comment]
Hi Mark,
Thanks for submitting the issue. A PR with a fix would be great, thanks!

**Comment by MarkBorazio** [comment]
I have opened a PR here: https://github.com/JuulLabs-OSS/mcumgr-ios/pull/52

**Comment by bgiori** [comment]
Closing issue, bugfix has been merged and released into 0.11.1

Thanks @MarkBorazio
